packages/circuits/test/proveGraffitiPreimage.test.ts

```diff
@@ -0,0 +1,147 @@
+    const GSTreeRoot = GSTree.root
+
+    const circuitInputs = {
+        epoch,
+        identity_nullifier: id.identityNullifier,
+        identity_trapdoor: id.trapdoor,
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+    }
+    return crypto.stringifyBigInts(circuitInputs)
+}
+
+const genProofAndPublicSignals = async (inputs: any): Promise<any> => {
+    const circuitWasmPath = path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.wasm`
+    )
+    const zkeyPath = path.join(__dirname, `../zksnarkBuild/${circuitName}.zkey`)
+    const { proof, publicSignals } = await snarkjs.groth16.fullProve(
+        inputs,
+        circuitWasmPath,
+        zkeyPath
+    )
+    return { proof, publicSignals }
+}
+
+const verifyProof = async (publicSignals, proof): Promise<boolean> => {
+    const vkey = require(path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.vkey.json`
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+describe('Prove grafitti preimage', function () {
+    this.timeout(300000)
+
+    const epoch = 1
+    const nonce = 1
+
+    it('successfully prove graffiti preimage', async () => {
+        const attesterId = 1
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(0),
+                BigInt(1),
+                BigInt(1)
+            ),
+        }
+        const circuitInputs = genCircuitInput(
+            new ZkIdentity(),
+            epoch,
+            reputationRecords,
+            attesterId
+        )
+
+        const { proof, publicSignals } = await genProofAndPublicSignals(
+            circuitInputs
+        )
+        const isValid = await verifyProof(publicSignals, proof)
+        expect(isValid).to.be.true
+    })
+
+    it('should fail to prove a reputation with wrong graffiti preimage', async () => {
+        const attesterId = 1
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(20),
+                BigInt(0),
+                BigInt(1)
+            ),
+        }
+        const id = new ZkIdentity()
+        const circuitInputs = genCircuitInput(
+            id,
+            epoch,
+            reputationRecords,
+            attesterId
+        )
+
+        const { proof, publicSignals } = await genProofAndPublicSignals(
+            circuitInputs
+        )
+        const isValid = await verifyProof(publicSignals, proof)
+        expect(isValid).to.be.true
```

Comment:
it should fail the test so you should use
```ts
expect(isValid).to.be.false
```
and you should also check the `hash(preimage) !== grafiiti`